site_tools/image_deployer.py

```diff
@@ -0,0 +1,316 @@
+                                                disk_format=disk_format,
+                                                container_format="bare",
+                                                visibility="private",
+                                                data=image_data,
+                                                **manifest_data)
+    logging.info(f"Uploaded image {new_image.name}.")
+    return new_image
+
+
+def archive_image(image_connection, image_name, new_image):
+    # TODO: if things go sideways in here we may be in a bad state. add more error handling
+    # rename the old image: this should only ever be 1. assert that instead?
+    existing_images = list(image_connection.image.images(name=image_name))
+    if len(existing_images) > 0:
+        logging.info(f"Renaming existing image {image_name}.")
+        archive_date = datetime.datetime.now().strftime("%Y%m%d_%H%M")
+        for existing_image in existing_images:
+            image_connection.image.update_image(existing_image.id, name=f"{image_name}_{archive_date}")
+            logging.info(f"Renamed image {existing_image.name} to {existing_image.name}_{archive_date}.")
+    image_connection.image.update_image(new_image.id, name=image_name)
+    logging.info(f"Renamed image {new_image.id} to {image_name}.")
+
+
+def sync_image(image,
+               object_connection,
+               image_connection,
+               current,
+               image_prefix="_testing",
+               dry_run=False):
+    logging.info(f"Syncing image {image.name}")
+    # TODO: move dry run to more of the steps
+    if dry_run:
+        logging.info(f"DRY RUN: Syncing image {image.name}.")
+    else:
+        logging.info(f"Syncing image {image.name}.")
+
+        logging.info(f"Downloading image {image.name} from {image.container_path}.")
+        manifest_object = object_connection.object_store.download_object (
+            container=image.container_path,
+            obj=image.manifest_name
+        )
+        manifest_data = json.loads(manifest_object.decode('utf-8'))
+        manifest_data["current"] = current
+        logging.debug(f"Downloaded {image.name} manifest: {manifest_data}, downloading image file.")
+
+        try:
+            # we need to use swift directly here because the images may be too large to fit
+            # in memory as they are downloaded and the swift client supports chunked downloads
+            swift_conn = setup_swift_connection(object_connection)
+            temp_raw_file = download_image_to_temp_file(
+                swift_conn,
+                image.base_container,
+                image.container_path,
+                image.raw_name
+            )
+            temp_qcow_file = download_image_to_temp_file(
+                swift_conn,
+                image.base_container,
+                image.container_path,
+                image.qcow_name
+            )
+            swift_conn.close()
```

Comment:
mentioned above as well, but ideally we should be able to fetch the image file with a non-opinionated http download. The context for this, besides not needing credentials to access the image bucket, is that we can later [enable Glance's "image import: web-download" feature](https://docs.openstack.org/glance/2023.1/admin/interoperable-image-import.html#configuring-the-web-download-method), and this tool can tell glance to fetch the image contents directly from the source URL, rather than needing to download+upload.

Reply:
done (i think)